blog/content/post/2021-12-13-secure-jtag-on-imx-soc.md

```diff
@@ -0,0 +1,280 @@
+---
+title: 'Increasing the security of iMX platforms - JTAG fusing'
+abstract: 'JTAG helps a lot of engineers during product development. It also may
+          be helpful for hackers. I tell you why and how to increase JTAG
+          security in your product'
+cover: /img/secure_jtag_cover.png
+author: michal.kotyla
+layout: post
+published: true
+date: 2021-12-13
+archives: "2021"
+
+tags:
+  - security
+  - jtag
+  - fusebits
+  - fusing
+  - nxp
+  - i.mx6
+categories:
+  - Security
+  - Manufacturing
+
+---
+## Introduction
+
+JTAG port is an important feature that helps a lot of engineers during product
+development. It also may be helpful for hackers while performing a
+reverse-engineering process - e.g. they can dump the memory of your firmware and
+get access to strictly confidential information. In this article, we show how to
+lock JTAG access for users who don't have a special key. We use for that Secure
+JTAG feature which is implemented on i.MX SoC's. In our lab, we tested this
+solution on i.MX6 but this same feature is available on a lot more NXP
+platforms. Unfortunately they have missing public documentation.
+
+## Requirements
+
+- **Development board**: Any devkit with i.MX6.
+
+- **SoC reference manual**: Setting JTAG in secure mode is a dangerous
+operation - if you miss something or will be wrong with any fusebit name or
+address you can enable unexcepted options. Here is an example: system is
+prepared to boot from SD card but you were wrong while fusemap understanding and
+now SoC always will be trying to boot from not existing eMMC memory. Read chip
+documentation and check twice that everything is correct.
+
+- **crucible tool**: JTAG mode and access key are saved in fuse bits - we need
+software to manipulate this in userspace.
+[crucible](https://github.com/f-secure-foundry/crucible/) tool can be used for
+that.
+
+- **JTAG compatible with OpenOCD**: On NXP sites we can read that Secure JTAG is
+supported only by the Lauterbach environment and ARM-DS5 IDE with DSTREAM
+debugger. In 3mdeb we are always trying to use open-source software.
+Unfortunately, OpenOCD doesn't officially support Secure JTAG in i.MX SoC's,
+but we found not accepted yet patch from 2014 which add support for work with
+SJC (System JTAG controller) in a secure mode. This patch work on OpenOCD from
+[official repository](git://git.code.sf.net/p/openocd/code) if you checkout this
+with `970a12aef` hash. There is a chance to apply this in the actual version of
+OpenOCD, but it needs some code changes. As a hardware, we used `ARM-USB-OCD-H`
+from Olimex.
```

Comment:
This whole procedure should be described step-by-step. More to that we should reply to the patch to indicate to the community that patch still works and is useful. Maybe even re-post the patch, so potentially someone else will base on most recent work.

Reply:
@kotylamichal so there are 2 things here:
1. building OpenOCD + patch - I think building without patch can be explained by pointing to relevant public documentation
2. reposting patch - if you plan to improve things for i.MX8 that's ok, but question if there is budget for that (@macpijan) and when it really can happen, improved version of patch that works for i.MX6 should be somehow given to community either through mailing list either through our repo of OpenOCD